### DISCUSSION TITLE
Modify column of stream

### DISCUSSION THREAD

**Discussion by ankitsoin** [finos/TimeBase-CE#228] [Q&A]
I am trying to add column in the stream as mentioned qql to modify stream in the documentation. I doing via python code.
My stream is having 25 spaces with around 10 millions records in each spaces.

When i ran modify qql via python code timebase got halted.
Now even if i do select * for any stream in timebase shell it is showing no response, just stucked.

May I know why this is happening and what could be the resolution to this?

I was using confirm drop_data even though i dont want to lose my data at all.
Previously when I ran modify qql command with confirm drop_data from timebase shell it was working fine for other stream. But when i ran via code it is stucked.

**Comment by ankitsoin** [comment]
Even why i try to restart timebase instance it is showing java.lang.RuntimeException: Failed to lock /mypath/timebase/tb.lock

**Comment by ankitsoin** [comment]
When i try to modify stream via timebase shell as well now i am not able to modify stream.
I just added MODIFY infront of what desc streamname shows.
But removing 'durable' key word and adding CONFIRM NO_CONVENSION at the end before \ 

It is shwoing me this exception as per snip
![20230426_005050](https://user-images.githubusercontent.com/22901224/234390834-0cd92933-c2e3-4973-a073-550c5cbe91c0.jpg)

**Reply by ankitsoin** [reply]
I am just adding new varchar column.
May I know, what this NO_CONVESION actually mean and is it even correct to use in my case.
Please suggest, thanks

I dont want to lose my spaces data.

**Reply by ankitsoin** [reply]
@alex-karpovich can you please provide me some resolution and answers to my issue.

**Reply by alex-karpovich** [reply]
please attach query you execute

**Reply by alex-karpovich** [reply]
Can you start TimeBase now?

**Comment by alex-karpovich** [comment]
Having sample stream like this
```
DURABLE STREAM TEST (
    CLASS "deltix.qsrv.hf.pub.BestBidOfferMessage" 'Quote Message' (
        "offerPrice" 'Offer Price' FLOAT DECIMAL (2),
        "offerSize" 'Offer Size' FLOAT DECIMAL (0),
        "bidPrice" 'Bid Price' FLOAT DECIMAL (2) RELATIVE TO "offerPrice",
        "bidSize" 'Bid Size' FLOAT DECIMAL (0)
    );
    CLASS "deltix.qsrv.hf.pub.TradeMessage" 'Trade Message' (
        "price" 'Trade Price' FLOAT DECIMAL (2),
        "size" 'Trade Size' FLOAT DECIMAL (0)
    );
)
OPTIONS (DF = 1; HIGHAVAILABILITY = FALSE)
COMMENT 'tickquerydemo'
```

best way to add columns with data rewrite - add them at the end and make it nullable.

```
MODIFY STREAM TEST (
    CLASS "deltix.timebase.api.messages.BestBidOfferMessage" 'Quote Message' (
        "offerPrice" 'Offer Price' FLOAT DECIMAL (2),
        "offerSize" 'Offer Size' FLOAT DECIMAL (0),
        "offerExchange" 'Offer Exchange' VARCHAR ALPHANUMERIC (10),
        "bidPrice" 'Bid Price' FLOAT DECIMAL (2) RELATIVE TO "offerPrice",
        "bidSize" 'Bid Size' FLOAT DECIMAL (0),

        "bidExchange" 'Bid Exchange' VARCHAR ALPHANUMERIC (10)
    );
    CLASS "deltix.timebase.api.messages.TradeMessage" 'Trade Message' (
        "price" 'Trade Price' FLOAT DECIMAL (2),
        "size" 'Trade Size' FLOAT DECIMAL (0)
    );
)
OPTIONS (DF = 1; HIGHAVAILABILITY = FALSE)
COMMENT 'Modified stream' CONFIRM DROP_DATA
```

**Reply by ankitsoin** [reply]
By adding column at the end worked, for my huge data in stream.
Thanks

I have 3 more queries
1.What does no_convension do and why we use it?
2.Why confirm data_drop is not dropping my datasets .even though it says drop data?
3. Do we have concept of sub spaces so that my product category stays at top heirarchy and respectives dates spaces under it?

**Reply by ankitsoin** [reply]
@alex-karpovich pls let me know above information?

**Reply by ankitsoin** [reply]
Hi @alex-karpovich  
Is the any method to rename the existing column name. I couldn't find in documentation anywhere.
I have added and removed column previously.

Secondly, I also want to know what does "RELATIVE TO" means? does it help in performance while fetching data from stream?
"bidPrice" 'Bid Price' FLOAT DECIMAL (2) RELATIVE TO "offerPrice",

**Reply by alex-karpovich** [reply]
To make clear:

`MODIFY STREAM` statement applies new schema provided, and you need just adjust previous schema, for example remove columns

https://kb.timebase.info/docs/development/modeling/annotations#relativeto-annotation
